code/net/ConnectX-5/mlx5.c

```diff
@@ -0,0 +1,412 @@
+  static_assert(UINTPTR_MAX <= UINT64_MAX, "uintptr_t must fit in an uint64_t");
+
+  // First make sure the PCI device is really an Mellanox ConnectX-5 Network Connection
+  uint32_t pci_id = pcireg_read(pci_address, PCIREG_ID);
+
+  if (pci_id != ((PCI_ID_HIGH << 16) | PCI_ID_LOW)) {
+    TN_DEBUG("PCI device is not what was expected pci_id %x", pci_id);
+    return false;
+  }
+
+  // TODO Check if connectx5 have these settings by default:
+
+  // Section 7.1:
+  // The driver should ensure that the Bus Master bit in the Command Register is set in the PCI
+  // configuration header
+  pcireg_set_field(pci_address, PCIREG_COMMAND, PCIREG_COMMAND_BUS_MASTER_ENABLE);
+//  // Same for memory reads, i.e. actually using the BARs.
+  pcireg_set_field(pci_address, PCIREG_COMMAND, PCIREG_COMMAND_MEMORY_ACCESS_ENABLE);
+//  // Finally, since we don't want interrupts and certainly not legacy ones, make sure they're disabled
+  pcireg_set_field(pci_address, PCIREG_COMMAND, PCIREG_COMMAND_INTERRUPT_DISABLE);
+
+
+  uint32_t pci_bar0low = pcireg_read(pci_address, PCIREG_BAR0_LOW);
+  // Sanity check: a 64-bit BAR must have bit 2 of low as 1 and bit 1 of low as 0 as per Table 9-4 Base Address Registers' Fields
+  if ((pci_bar0low & BIT(2)) == 0 || (pci_bar0low & BIT(1)) != 0) {
+    TN_DEBUG("BAR0 is not a 64-bit BAR");
+    return false;
+  }
+  uint32_t pci_bar0high = pcireg_read(pci_address, PCIREG_BAR0_HIGH);
+  // No need to detect the size, since we know exactly which device we're dealing with. (This also means no writes to BARs, one less chance to mess everything up)
+  // BARs ==  actual pointer
+
+  struct tn_net_device device = { 0 };
+  // Section 9.3.6.1 Memory and IO Base Address Registers:
+  // As indicated in Table 9-4, the low 4 bits are read-only and not part of the address
+  uintptr_t dev_phys_addr = (((uint64_t) pci_bar0high) << 32) | (pci_bar0low & ~BITS(0,3));
+  // Section 8.1 Address Regions: "Region Size" of "Internal registers memories and Flash (memory BAR)" is "128 KB + Flash_Size"
+  // Thus we can ask for 128KB, since we don't know the flash size (and don't need it thus no need to actually check it)
+  if (!tn_mem_phys_to_virt(dev_phys_addr, 128 * 1024, &device.addr)) {
+    return false;
+  }
+
+  TN_VERBOSE("Device %02" PRIx8 ":%02" PRIx8 ".%" PRIx8 " mapped to %p", pci_address.bus, pci_address.device, pci_address.function, device.addr);
+
+  /**
+   * Step 1.
+   * Read the initialization segment from offset 0 of the HCA BAR,
+   * to retrieve the versions of the firmware and the command interface.
+   * The driver must match the command interface revision number.
+   * The format of the initialization segment is in Table 12, “Initialization Segment,” on page 168
+   * */
+  TN_VERBOSE("### Init - step 1");
+
+  uint32_t fw_rev = reg_read(device.addr, 0x00);
+  uint32_t cmd_rev_fw_rev_subminor = reg_read(device.addr, 0x04);
+  uint16_t fw_rev_minor = fw_rev >> 16;
```

Comment:
You can read this kind of stuff with the utility functions above, e.g. `reg_read_field(device.addr, 0x04, BITS(16,31))` will get you that fw_rev_minor without having to do the shifts and masks yourself.